### PR TITLE
Add batches for test data predicts.

### PR DESCRIPTION
**Pull request by petrroll**
Predicting all data at once OOM'ed on fashion_masks task with larger networks. This fixes it (I can provide similar PR for fashion_mask as well). 

BTW: I have a few other quality-of-life improvements such as explicit file_name arg. that gets prefixed to the predicted result file etc. Should I create PRs for them as well or are the templates purposefully barebones?

### CODE REVIEW

File: labs/06/3d_recognition.py [hunk 1]
```diff
@@ -3,12 +3,13 @@
...
 
 class Dataset:
-    def __init__(self, filename):
+    def __init__(self, filename, shuffle_batches = True):
         data = np.load(filename)
         self._voxels = data["voxels"]
         self._labels = data["labels"] if "labels" in data else None
 
-        self._permutation = np.random.permutation(len(self._voxels))
+        self._shuffle = shuffle_batches
```
Comment (foxik) [inline labs/06/3d_recognition.py]: Could you please rename it to _shuffle_batches, mostly for consistency with the argument name?

File: labs/06/3d_recognition.py [hunk 2]
```diff
@@ -36,9 +37,14 @@ def next_batch(self, batch_size):
         batch_perm, self._permutation = self._permutation[:batch_size], self._permutation[batch_size:]
         return self._voxels[batch_perm], self._labels[batch_perm]
 
+    def next_batch_data_for_prediction(self, batch_size):
```
Comment (foxik) [inline labs/06/3d_recognition.py]: Please get rid of the next_batch_data_for_prediction; instead, make `next_batch` return None for labels, i.e., something like
```python
  return self._voxels[batch_perm], self._labels[batch_perm] if self._labels is not None else None
```

**Review by foxik — CHANGES_REQUESTED** [review]

**Comment by foxik** [comment]
I am running on CPU, so I do not mind the large required memory :-)

Thanks for the pull request, I have two comments in the review, if you could address them; if not, I will do them myself anyway :-)

To file names: I assumed students would use some early stopping/ensembling/..., so they would be generating test set annotation/distributions for each epoch anyway, so I am not sure if it is worth it; but I am definitely not opposed to it, so feel free to send a pull request (with argument called probably test_annotations_file?).

Pull requests for fashion_masks are welcome ;-)

**Comment by petrroll** [comment]
Should be ok now.

**Comment by petrroll** [comment]
Should be ready to merge :).

**Comment by petrroll** [comment]
I have the other proposed changes in https://github.com/petrroll/npfl114/tree/more-options. Will create new PR once this gets merged.

**Comment by foxik** [comment]
Thanks, merging.

As for the proposed changes in the more-options branch:
- please do not use literal string interpolation, it needs Python 3.6
- I am not much in favor of adding predict_test -- compared to for example batching of test prediction, it only adds trivial functionality, it will "clutter" the logdirs and (totally subjective) feels unnecessary
- the run name is encoded in the logdir automatically as `rn=name`, so there is no need of explicitly adding it to args.logdir; from this point of view it would make sense to just have --run (encoded as r=name)
- the default value of run should be an empty string (and the hyphen before it [for example in test set annotation name] should be added only when needed)

**Comment by petrroll** [comment]
You're probably right that those additions are too small and not general
enough for them to be justifiable additions to  the templates.

**Comment by foxik** [comment]
Adding the name of the run could be useful -- I think I will add it anyway :-)

**Comment by petrroll** [comment]
BTW: Speaking of run_name being appended to the args.logdir. There's an issue with the logdir name when the arguments are a bit longer (I suspect the path can't be longer than 256 chars). Which can - surprisingly - happen quite easily. 

I have a local fix that omits the appended arguments and uses run_name instead. Hence I added it explicitly (copying code always leads to errors :D)

**Comment by martinpopel** [comment]
BTW: I think most filesystems have no **path** length limit, some systems have 4096. However, most filesystems have **filename** length limit 255 (usually bytes, not chars). In my experiments I use logdirs with a special directory level for each hyperparameter, e.g. `batch_size1000/learning_rate2/...`. The advantage is that I can run TensorBoard with just a subset of all runs (in case the subset corresponds to a directory subtree, otherwise I have to create an alternative hierarchy with symlinks). This is useful if the tensorboard takes several GB of RAM and loads slowly with all the runs. (I know this is a bit off-topic in case of npfl117 hw.)

**Comment by foxik** [comment]
It is true that logdir can be too long; that is why I am shortening the arguments. If the logdir is too long, it is also not visible correctly in TensorBoard popups.

However, I used the "just supply a run name" approach before, but it has two issues:
- when running a lot of experiments, it is difficult to pass consistent run_names, so in the end I was not sure which experiment is which (especially if I work on a task for several weeks)
- many times I wanted to know exact hyperparameters of a run, but had only a run_name.

The second issue could be solved using the text summaries, but they are not (directly) available in the V2 summary API yet.

@petrrol I am thinking about generating the test set annotation in the args.logdir. This way it will be easy to match test set annotations to runs. Thoughts?

**Comment by petrroll** [comment]
That sounds like a great idea.

**Comment by foxik** [comment]
Ok, added in 34f0241.
